Feature/rove comm tcp

Joshua-Lawson opened the pull request:
Implements RoveCommTcp, and ties instances of RoveCommUdp and RoveCommTcp to a single listener thread

Commit:
Added TCP to RoveComm and made all instances of RoveComm listen on the same thread

Commit:
Removed unnecessary second Tcp instance

Commit:
Merge branch 'dev' into feature/RoveCommTCP

BnetButter commented on core/rovecomm.py:
```diff
@@ -141,46 +182,157 @@ def read(self):
-            packet, remote_ip = self.RoveCommSocket.recvfrom(1024)
-            header_size = struct.calcsize(ROVECOMM_HEADER_FORMAT)
+        ready = select.select([self.RoveCommSocket], [], [], 0)
+        if len(ready[0]) != 0:
```
You can just do `if read[0]:`

thomas-gleiforst replied:
If ready[0] would imply we are looking to see if the element at index 0 evaluates to true, where here we are looking if ready[0] exists/has length. You could just say len(ready[0]), but != 0 is more transparent

BnetButter commented on core/rovecomm.py:
```diff
@@ -73,6 +86,34 @@ def print(self):
+
+    def __init__(self):
+        self.thread = threading.Thread(target=self.listen)
+        self.thread.daemon = True
```
If the program terminates abnormally, the daemon thread may terminate without calling cleanup function

Yajirushi12 replied:
This is a good point, but it might not be as important in this case as in some other cases. If the program terminates abnormally we have a bigger issue than threads not cleaning up.

Commit:
Changed rovecomm thread to non-daemon

Listener thread now closes after the main thread, or when the shutdown event is communicated to it

EliVerbrugge commented on core/rovecomm.py:
```diff
@@ -141,46 +187,145 @@ def read(self):
+        for open_socket in available_sockets[0]:
+            if open_socket is self.server:
+                conn, addr = self.server.accept()
+                self.open_sockets[addr[0]] = conn
```
add a new function to handle incoming connections, call that from the RoveComm thread.

EliVerbrugge commented on core/rovecomm.py:
```diff
@@ -141,46 +187,145 @@ def read(self):
+        return packets
+
+
+collection = None
```
probably move this to the top of the file

EliVerbrugge commented on core/rovecomm.py:
```diff
@@ -141,46 +187,145 @@ def read(self):
-        '''
-        while True:
-            packet = self.read()
+    def connect(self, address):
```
For future, we're going to be working on adding docstrings to our functions. https://www.programiz.com/python-programming/docstrings

EliVerbrugge commented on rovecomm_test.py:
```diff
@@ -0,0 +1,44 @@
+    RoveCommTCP = RoveCommEthernetTcp(HOST='127.0.0.1', PORT=11111)
+    RoveCommTCP.callbacks[4242] = addCounter
+    packet2 = RoveCommPacket(4242, 'b', (1, 3), "", 11111)
+    packet2.SetIp('127.0.0.1')
```
Either create two RoveCommTCP instances, or clarify as a comment that write() creates a new socket, which in this case connects back to RoveCommTCPs server socket. Not super evident in this example.

EliVerbrugge commented on rovecomm_test.py:
```diff
@@ -0,0 +1,44 @@
+rovecomm_test.variable = 0
+
+
+def main() -> None:
```
Will be converting this to a full on unit test.

EliVerbrugge commented on core/rovecomm.py:
```diff
@@ -73,6 +86,39 @@ def print(self):
+        self.thread.start()
+
+    def listen(self):
+        while threading.main_thread().is_alive() and not self.shutdown_event.isSet():
```
Confirm 100% that this is the correct way of checking whether we should still be running.

EliVerbrugge commented on core/rovecomm.py:
```diff
@@ -73,6 +86,39 @@ def print(self):
         print('----------')
 
 
+class RoveCommCollection:
```
Call this just RoveComm, and this should initialize a RoveCommUDP and RoveCommTCP instance. Also define a write function that takes a boolean "Reliable" parameter that determines if it should send over UDP and TCP.

EliVerbrugge commented on core/rovecomm.py:
```diff
@@ -73,6 +86,39 @@ def print(self):
+        for connection in self.connections:
+            if isinstance(connection, RoveCommEthernetTcp):
+                connection.close_sockets()
+
```
Create addEvent() or something similar, to add a function and data ID to the callbacks dictionary.

EliVerbrugge commented on core/rovecomm.py:
```diff
@@ -141,46 +187,145 @@ def read(self):
+        for key, value in self.open_sockets.items():
+            available_sockets.append(value)
+
+        available_sockets = select.select(available_sockets, [], [], 0)
```
Add a brief comment to describe this

Commit:
Changed the collection to initialize its children

Commit:
Revert "Changed the collection to initialize its children"

This reverts commit fe8e6e05cde75cfd967eec52a9f3bfe18ed258bb.

Commit:
Changed RoveCommCollection to initialize both protocols

Commit:
added handle_incoming_connection function

Commit:
Removed semaphore and explained select function

Commit:
Added some docstrings

Commit:
More docstrings

Commit:
Improved class docstrings

Commit:
Added a unit test

Commit:
Got to 100% coverage for core/rovecomm.py

Commit:
Added some comments to unit test functions

EliVerbrugge commented on tests/unit/rovecomm_test.py:
```diff
@@ -0,0 +1,306 @@
+    assert responses.get(4233, (1, 2, 3)) == (1, 2, 3)
+
+
+def test_old_version_tcp():
```
rename these to test_invalid_rovecomm_version_tcp() and udp() or something, took me a second to catch what was being tested.

EliVerbrugge commented on tests/unit/rovecomm_test.py:
```diff
@@ -0,0 +1,306 @@
+
+    # Packet should not still be recieved because we unsubscribed
+    time.sleep(.05)
+    assert responses.get(4233, (1, 2, 3)) == (1, 2, 3)
```
Minor nitpick, probably change to get(4233) == None, a tiny bit ambiguous right now that we are actually specifying the return value as (1,2,3) if there is nothing there.

EliVerbrugge commented on tests/unit/rovecomm_test.py:
```diff
@@ -0,0 +1,306 @@
+    core.rovecomm.udp_node.RoveCommSocket = temp
+
+
+def test_listener_shutdown():
```
hmmm, this will cause issues in other unit tests if we rely on core.rovecomm having a listening thread. It should be done, we will just need to adjust future tests accordingly.

EliVerbrugge commented on core/rovecomm.py:
```diff
@@ -141,46 +277,187 @@ def read(self):
+        '''
+        Closes the UDP socket
+        '''
+        self.RoveCommSocket.shutdown(1)
```
We shouldn't be calling shutdown() on a DGRAM socket. I don't think it has defined behaviour. Shouldn't be a problem to take out.

EliVerbrugge commented on core/rovecomm.py:
```diff
@@ -141,46 +277,187 @@ def read(self):
+        self.open_sockets = {}
+        # configure a TCP socket
+        self.server = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+        # bind the socket to the current machines local network IP by default (can be specified as well)
```
should add the following line: self.server.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1). This will allow us to reopen the socket instantly after we close it, allowing us to restart Autonomy code immediately.

EliVerbrugge replied:
Add this before the bind() call

EliVerbrugge reviewed (approved):
Looks good and is functional, got a couple of minor comments you should address before we merge. Mostly just fixes that will allow us to rerun autonomy code without waiting for sockets to timeout.

Commit:
TCP socket now releases the port when finished
